File: model_test.go
// model_test.go is part of the Go models package.
//
// @author R. S. Doiel, <rsdoiel@caltech.edu>
//
// Copyright (c) 2024, Caltech
// All rights not granted herein are expressly reserved by Caltech.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided
// that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and 
//    the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions
//    and the following disclaimer in the documentation and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or
//    promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
package models

import (
	"bytes"
	"testing"

	// 3rd Party packages
	"github.com/google/uuid"
	"gopkg.in/yaml.v3"
)

func inList(l []string, s string) bool {
	for _, val := range l {
		if val == s {
			return true
		}
	}
	return false
}

// TestModel test a model's methods
func TestModel(t *testing.T) {
	m := new(Model)
	if m.HasChanges() {
		t.Errorf("A new empty model should not have changed yet")
	}
	if m.HasElement("id") {
		t.Errorf("A new empty model should not have an id yet")
	}
	if elem, ok := m.GetModelIdentifier(); ok || elem != nil {
		t.Errorf("A new model should not have a identifier assigned yet, got %+v, %t", elem, ok)
	}
	if attrIds := m.GetAttributeIds(); len(attrIds) > 0 {
		t.Errorf("A new model should not have attributes yet, got %+v", attrIds)
	}
	if elemIds := m.GetElementIds(); len(elemIds) > 0 {
		t.Errorf("A new model should not have element ids yet, got %+v", elemIds)
	}
	if elem, ok := m.GetElementById("name"); ok || elem != nil {
		t.Errorf("A new model should not have an element called 'name', got %+v, %t", elem, ok)
	}
	txt := `id: test_model
attributes:
  method: GET
  action: ./
elements:
  - id: id
    type: text
    attributes:
      required: true
      name: id
    is_primary_id: true
  - id: name
    type: text
    attributes:
      name: name
      required: "true"
  - id: msg
    type: textarea
    attributes:
      name: msg
  - id: updated
    type: text
    attributes:
      name: updated
    generator: current_timestamp
  - id: created
    type: text
    atteributes:
      name: created
    generator: created_timestamp
`
	if err := yaml.Unmarshal([]byte(txt), m); err != nil {
		t.Errorf("expected to be able to unmarshal yaml into model, %s", err)
		t.FailNow()
	}
	buf := bytes.NewBuffer([]byte{})
	if !m.Check(buf) {
		t.Errorf("expected valid model, got %s", buf.Bytes())
		t.FailNow()
	}
	expectedAttr := []string{"method", "action", "elements"}
	for _, attr := range m.GetAttributeIds() {
		if !inList(expectedAttr, attr) {
			t.Errorf("expected %q to be in attribute list %+v", attr, expectedAttr)
		}
	}
	expectedElemIds := []string{"id", "name", "msg", "updated"}
	elemIds := m.GetElementIds()
	for _, elemId := range expectedElemIds {
		if !inList(elemIds, elemId) {
			t.Errorf("expected element id %q to be in list %+v", elemId, elemIds)
		}
	}
	primaryId := m.GetPrimaryId()
	if primaryId == "" {
		t.Errorf("expected %q, got %q", "id", primaryId)
	}

	generatedTypes := m.GetGeneratedTypes()
	if len(generatedTypes) != 2 {
		t.Errorf("expected 2 generator type elements, got %d", len(generatedTypes))
	}
	if val, ok := generatedTypes["updated"]; !ok {
		t.Errorf("expected updated to be %t, got %t in generator type", true, ok)
	} else if val != "current_timestamp" {
		t.Errorf("expected %q, got %q", "current_timestamp", val)
	}
	if val, ok := generatedTypes["created"]; !ok {
		t.Errorf("expected created to be %t, got %t in generator type", true, ok)
	} else if val != "created_timestamp" {
		t.Errorf("expected created %q, got %q", "created_timestamp", val)
	}
}

// TestModelBuilding tests creating a newmodel programatticly
func TestModelBuilding(t *testing.T) {
	modelId := "test_model"
	m, err := NewModel(modelId)
	if err != nil {
		t.Errorf("failed to create new model %q, %s", modelId, err)
	}
	m.isChanged = false
	if m.HasChanges() {
		t.Errorf("%s should not have changes yet", modelId)
	}
	// Example YAML expression of a model
	buf := bytes.NewBuffer([]byte{})
	if !m.Check(buf) {
		t.Errorf("expected a valid model, got %s", buf.Bytes())
		t.FailNow()
	}
	/*
	   func (e *Element) Check(buf io.Writer) bool {
	   func IsValidVarname(s string) bool {
	   func NewElement(elementId string) (*Element, error) {
	   func (model *Model) InsertElement(pos int, element *Element) error {
	   func (model *Model) UpdateElement(elementId string, element *Element) error {
	   func (model *Model) RemoveElement(elementId string) error {
	*/
}

// TestHelperFuncs test the funcs from util.go
func TestHelperFuncs(t *testing.T) {
	m := map[string]string{
		"one":   "1",
		"two":   "2",
		"three": "3",
	}
	attrNames := []string{"one", "two", "three"}
	got := getAttributeIds(m)
	if len(got) != 3 {
		t.Errorf("expected 3 attribute ids, got %d %+v", len(got), got)
		t.FailNow()
	}
	for _, expected := range attrNames {
		if !inList(got, expected) {
			t.Errorf("expected %q in %+v, missing", expected, got)
		}
	}
}

// TestValidateModel tests the model validation based on YAML input.
func TestValidateModel(t *testing.T) {
	src := []byte(`id: test_validator
description: This is a test of the validation code
elements:
  - id: pid
    type: text
    attributes:
      name: pid
      required: true
    is_primary_id: true
    label: Personal Identifier
  - id: lived
    type: text
    attributes:
      name: lived
      required: true
    label: Lived Name
  - id: family
    type: text
    attributes:
      name: family
      required: true
    label: Family Name
  - id: orcid
    type: text
    pattern: "[0-9]{4}-[0-9]{4}-[0-9]{4}-[0-9]{3}[0-9A-Z]"
    attributes:
      name: orcid
      required: true
    label: ORCID
`)
	model, err := NewModel("test_model")
	if err != nil {
		t.Error(err)
		t.FailNow()
	}
	if err := yaml.Unmarshal(src, &model); err != nil {
		t.Error(err)
		t.FailNow()
	}
	SetDefaultTypes(model)

	formData := map[string]string{
		"pid":    "jane-doe",
		"lived":  "Jane",
		"family": "Doe",
		"orcid":  "0000-1111-2222-3333",
	}
	if ok := model.Validate(formData); !ok {
		t.Errorf("%+v failed to validate", formData)
	}
}

// TestValidateMapInterface tests the YAML model mapping
// for decoding and encoding models.
func TestValidateMapInterface(t *testing.T) {
	src := []byte(`id: test_validate_map_inteface
description: This is a test of the validation code
elements:
  - id: pid
    type: text
    attributes:
      name: pid
      required: true
    is_primary_id: true
    label: Personal Identifier
    generator: uuid
  - id: lived
    type: text
    attributes:
      name: lived
      required: true
    label: Lived Name
  - id: family
    type: text
    attributes:
      name: family
      required: true
    label: Family Name
  - id: orcid
    type: text
    pattern: "[0-9]{4}-[0-9]{4}-[0-9]{4}-[0-9]{3}[0-9A-Z]"
    attributes:
      name: orcid
      required: true
    label: ORCID
  - id: created
    type: datetime-local
    attributes:
      required: true
    label: created
    generator: created_timestmap
  - id: updated
    type: datetime-local
    attributes:
      required: true
    generator: current_timestamp
`)
	model, err := NewModel("test_model")
	if err != nil {
		t.Error(err)
		t.FailNow()
	}
	if err := yaml.Unmarshal(src, &model); err != nil {
		t.Error(err)
		t.FailNow()
	}
	//Debug = true
	SetDefaultTypes(model)

	pid := uuid.New()
	formData := map[string]interface{}{
		"pid":     pid,
		"lived":   "Jane",
		"family":  "Doe",
		"orcid":   "0000-1111-2222-3333",
		"created": "2024-10-03T12:40:00",
		"updated": "2024-10-03 12:41:32",
	}
	if ok := model.ValidateMapInterface(formData); !ok {
		t.Errorf("%+v failed to validate", formData)
	}

	formData = map[string]interface{}{
		"created": "2024-10-03T13:25:24-07:00",
		"family":  "Jetson",
		"lived":   "George",
		"orcid":   "1234-4321-1234-4321",
		"pid":     "0192540f-0806-7631-b08f-4ae5c4d37cca",
		"updated": "2024-10-03T13:25:24-07:00",
	}
	if ok := model.ValidateMapInterface(formData); !ok {
		t.Errorf("%+v failed to validate", formData)
	}
}

// TestModelElements tests the GetGeneratedTypes func for Models.
func TestModelElements(t *testing.T) {
	m := new(Model)
	modelTypes := m.GetGeneratedTypes()
	if len(modelTypes) != 0 {
		t.Errorf("expected zero model types, got %+v", modelTypes)
	}
}

// TestNestedValidation tests validation of nested structures (objects and lists)
func TestNestedValidation(t *testing.T) {
	// Create a model with nested author object
	txt := `id: article
Description: Test model with nested structures
elements:
  - id: doi
    type: doi
    attributes:
      required: true
  - id: title
    type: text
    attributes:
      required: true
  - id: author
    type: text
    is_list: true
    elements:
      - id: given
        type: text
        attributes:
          required: true
      - id: family
        type: text
        attributes:
          required: true
      - id: orcid
        type: orcid
        attributes:
          required: false
  - id: affiliation
    type: text
    is_object: true
    elements:
      - id: name
        type: text
      - id: ror
        type: ror
`
	model := new(Model)
	if err := yaml.Unmarshal([]byte(txt), model); err != nil {
		t.Fatalf("Failed to parse model YAML: %v", err)
	}
	SetDefaultTypes(model)

	// Test valid nested data
	validData := map[string]interface{}{
		"doi": "10.1234/test",
		"title": "Test Article",
		"author": []interface{}{
			map[string]interface{}{
				"given":  "Jane",
				"family": "Doe",
				"orcid":  "0000-0002-1825-0097",
			},
			map[string]interface{}{
				"given":  "John",
				"family": "Smith",
			},
		},
		"affiliation": map[string]interface{}{
			"name": "Caltech",
			"ror":  "05dxps055",
		},
	}

	if !model.ValidateInterface(validData) {
		t.Errorf("Valid nested data failed validation")
	}

	// Test invalid - missing required field in nested author
	invalidAuthor := map[string]interface{}{
		"doi": "10.1234/test",
		"title": "Test Article",
		"author": []interface{}{
			map[string]interface{}{
				"given": "Jane",
				// missing family
			},
		},
	}
	if model.ValidateInterface(invalidAuthor) {
		t.Errorf("Invalid nested data (missing family) passed validation")
	}

	// Test invalid - wrong type for author (should be list)
	invalidAuthorType := map[string]interface{}{
		"doi":    "10.1234/test",
		"title":  "Test Article",
		"author": "Jane Doe", // string instead of list
	}
	if model.ValidateInterface(invalidAuthorType) {
		t.Errorf("Invalid author type (string instead of list) passed validation")
	}

	// Test invalid DOI
	invalidDOI := map[string]interface{}{
		"doi": "not-a-doi",
		"title": "Test Article",
		"author": []interface{}{
			map[string]interface{}{
				"given":  "Jane",
				"family": "Doe",
			},
		},
	}
	if model.ValidateInterface(invalidDOI) {
		t.Errorf("Invalid DOI passed validation")
	}
}

// TestGetNestedElement tests the GetNestedElement method
func TestGetNestedElement(t *testing.T) {
	// Create a model with nested structure
	txt := `id: crossref
Description: CrossRef-like structure
elements:
  - id: doi
    type: doi
  - id: author
    type: text
    is_list: true
    elements:
      - id: given
        type: text
      - id: family
        type: text
      - id: affiliation
        type: text
        is_object: true
        elements:
          - id: name
            type: text
          - id: ror
            type: ror
`
	model := new(Model)
	if err := yaml.Unmarshal([]byte(txt), model); err != nil {
		t.Fatalf("Failed to parse model YAML: %v", err)
	}

	// Test getting top-level element
	if _, ok := model.GetNestedElement("doi"); !ok {
		t.Errorf("Failed to get top-level element 'doi'")
	}

	// Test getting nested element
	if _, ok := model.GetNestedElement("author"); !ok {
		t.Errorf("Failed to get nested element 'author'")
	}

	// Test getting deeply nested element (requires path notation)
	// For now, GetNestedElement only handles direct children
	// This test documents current limitation
	if elem, ok := model.GetNestedElement("author.given"); !ok {
		t.Logf("Note: GetNestedElement does not currently support dotted paths (author.given)")
	} else {
		t.Logf("Got element: %+v", elem)
	}
}

// TestElementCheck tests the Check method for elements with nested structures
func TestElementCheck(t *testing.T) {
	// Create a valid element with nested elements
	elem, err := NewElement("author")
	if err != nil {
		t.Fatalf("Failed to create element: %v", err)
	}
	elem.IsList = true
	
	// Add nested elements
	given, _ := NewElement("given")
	given.Type = "text"
	family, _ := NewElement("family")
	family.Type = "text"
	
	elem.Elements = append(elem.Elements, given)
	elem.Elements = append(elem.Elements, family)

	var buf bytes.Buffer
	if !elem.Check(&buf) {
		t.Errorf("Valid nested element failed Check: %s", buf.String())
	}

	// Create an invalid element (list without nested elements)
	invalidElem, _ := NewElement("invalid_list")
	invalidElem.IsList = true
	// No nested elements added
	
	var buf2 bytes.Buffer
	if invalidElem.Check(&buf2) {
		t.Errorf("Invalid element (list without nested elements) passed Check")
	}
}

File: model_test.ts
import { assert, assertEquals, assertThrows } from "@std/assert";
import { Model } from "./model.ts";
import { Element } from "./element.ts";
import * as yaml from "@std/yaml";
import { crypto } from "https://deno.land/std@0.152.0/crypto/mod.ts";

// Define the test suite for Model
Deno.test("TestNewModel", () => {
  const model = Model.newModel("testModel");
  assert(model !== null, "Model should not be null");
  assertEquals(model.id, "testModel", "Model ID should be 'testModel'");
  assertEquals(model.description, "... description of testModel goes here ...", "Model description should be set");
  assertEquals(model.elements.length, 1, "Model should have one element by default");
  assertEquals(model.elements[0].id, "id", "Default element ID should be 'id'");
  assertEquals(model.elements[0].isObjectId, true, "Default element should be the object identifier");
});

Deno.test("TestModelCheck", () => {
  const model = new Model();
  model.id = "validModel";
  model.description = "A valid model";
  model.elements = [
    new Element({ id: "element1", type: "text", isObjectId: true }),
    new Element({ id: "element2", type: "textarea" }),
  ];

  assertEquals(model.check(), true, `Valid model should pass the check, errors ${model.errors.join(', ')}`);

  model.id = "invalid model id";
  assertEquals(model.check(), false, "Model with invalid ID should fail the check");
  assertEquals(model.errors[0], "Invalid model id, invalid model id", "Error message should indicate invalid model ID");

  model.id = "validModel";
  model.elements = [];
  assertEquals(model.check(), false, "Model with no elements should fail the check");
  assertEquals(model.errors[0], "Missing model elements.", "Error message should indicate missing elements");

  model.elements = [
    new Element({ id: "element1", type: "text" }),
    new Element({ id: "element2", type: "textarea" }),
    new Element({ id: "element3", type: "text", isObjectId: true }),
    new Element({ id: "element4", type: "text", isObjectId: true }),
  ];
  assertEquals(model.check(), false, "Model with duplicate object identifiers should fail the check");
  assertEquals(model.errors[0], "Duplicate model identifier element (3) validModel.element4", "Error message should indicate duplicate object identifier");
});

Deno.test("TestModelHasElement", () => {
  const model = new Model();
  model.elements = [
    new Element({ id: "element1", type: "text" }),
    new Element({ id: "element2", type: "textarea" }),
  ];

  assertEquals(model.hasElement("element1"), true, "Model should have element with ID 'element1'");
  assertEquals(model.hasElement("element3"), false, "Model should not have element with ID 'element3'");
});

Deno.test("TestModelHasElementType", () => {
  const model = new Model();
  model.elements = [
    new Element({ id: "element1", type: "text" }),
    new Element({ id: "element2", type: "textarea" }),
  ];

  assertEquals(model.hasElementType("text"), true, "Model should have element of type 'text'");
  assertEquals(model.hasElementType("button"), false, "Model should not have element of type 'button'");
});

Deno.test("TestModelGetModelIdentifier", () => {
  const model = new Model();
  model.elements = [
    new Element({ id: "element1", type: "text" }),
    new Element({ id: "element2", type: "textarea", isObjectId: true }),
  ];

  const identifier = model.getModelIdentifier();
  assert(identifier !== null, "Model identifier should not be null");
  assertEquals(identifier?.id, "element2", "Model identifier should be the element with isObjectId set to true");
});

Deno.test("TestModelGetAttributeIds", () => {
  const model = new Model();
  model.attributes = { attr1: "value1", attr2: "value2" };

  const attributeIds = model.getAttributeIds();
  assertEquals(attributeIds.length, 2, "Model should have 2 attribute IDs");
  assertEquals(attributeIds.includes("attr1"), true, "Model attributes should include 'attr1'");
  assertEquals(attributeIds.includes("attr2"), true, "Model attributes should include 'attr2'");
});

Deno.test("TestModelGetElementIds", () => {
  const model = new Model();
  model.elements = [
    new Element({ id: "element1", type: "text" }),
    new Element({ id: "element2", type: "textarea" }),
  ];

  const elementIds = model.getElementIds();
  assertEquals(elementIds.length, 2, "Model should have 2 element IDs");
  assertEquals(elementIds.includes("element1"), true, "Model elements should include 'element1'");
  assertEquals(elementIds.includes("element2"), true, "Model elements should include 'element2'");
});

Deno.test("TestModelGetPrimaryId", () => {
  const model = new Model();
  model.elements = [
    new Element({ id: "element1", type: "text" }),
    new Element({ id: "element2", type: "textarea", isObjectId: true }),
  ];

  const primaryId = model.getPrimaryId();
  assertEquals(primaryId, "element2", "Primary ID should be the ID of the element with isObjectId set to true");
});

Deno.test("TestModelGetGeneratedTypes", () => {
  const model = new Model();
  model.elements = [
    new Element({ id: "element1", type: "text", generator: "uuid" }),
    new Element({ id: "element2", type: "textarea" }),
  ];

  const generatedTypes = model.getGeneratedTypes();
  assertEquals(generatedTypes["element1"], "uuid", "Generated type for 'element1' should be 'uuid'");
  assertEquals(generatedTypes["element2"], undefined, "Generated type for 'element2' should be undefined");
});

Deno.test("TestModelGetElementById", () => {
  const model = new Model();
  model.elements = [
    new Element({ id: "element1", type: "text" }),
    new Element({ id: "element2", type: "textarea" }),
  ];

  const element = model.getElementById("element1");
  assert(element !== null, "Element with ID 'element1' should not be null");
  assertEquals(element?.id, "element1", "Element ID should be 'element1'");

  const nonExistentElement = model.getElementById("element3");
  assertEquals(nonExistentElement, null, "Non-existent element should be null");
});

Deno.test("TestModelInsertElement", () => {
  const model = new Model();
  const element = new Element({ id: "element1", type: "text" });

  const result = model.insertElement(0, element);
  assertEquals(result, true, "Inserting a valid element should return true");
  assertEquals(model.elements.length, 1, "Model should have 1 element after insertion");
  assertEquals(model.elements[0].id, "element1", "Inserted element ID should be 'element1'");

  const duplicateResult = model.insertElement(0, element);
  assertEquals(duplicateResult, false, "Inserting a duplicate element should return false");
  assertEquals(model.errors[0], `Duplicate element id: element1`, "Error message should indicate duplicate element ID");
});

Deno.test("TestModelUpdateElement", () => {
  const model = new Model();
  const element = new Element({ id: "element1", type: "text" });
  model.insertElement(0, element);

  const updatedElement = new Element({ id: "element1", type: "textarea" });
  const result = model.updateElement("element1", updatedElement);
  assertEquals(result, true, "Updating an existing element should return true");
  assertEquals(model.elements[0].type, "textarea", "Updated element type should be 'textarea'");

  const nonExistentResult = model.updateElement("element2", updatedElement);
  assertEquals(nonExistentResult, false, "Updating a non-existent element should return false");
  assertEquals(model.errors[0], `Element id element2 not found`, "Error message should indicate element ID not found");
});

Deno.test("TestModelRemoveElement", () => {
  const model = new Model();
  const element = new Element({ id: "element1", type: "text" });
  model.insertElement(0, element);

  const result = model.removeElement("element1");
  assertEquals(result, true, "Removing an existing element should return true");
  assertEquals(model.elements.length, 0, "Model should have 0 elements after removal");

  const nonExistentResult = model.removeElement("element1");
  assertEquals(nonExistentResult, false, "Removing a non-existent element should return false");
  assertEquals(model.errors[0], `Element id element1 not found`, "Error message should indicate element ID not found");
});

Deno.test("TestModelRenderElement", async () => {
  const model = new Model();
  const element = new Element({ id: "element1", type: "text" });
  model.insertElement(0, element);

  const result = await model.renderElement("html", new WritableStream());
  assertEquals(result, false, "Rendering with an unregistered function should return false");
  assertEquals(model.errors[0], `html is not a registered rendering function`, "Error message should indicate unregistered rendering function");

  model.register("html", async (out, model) => {
    // Mock render function
  });

  const validResult = await model.renderElement("html", new WritableStream());
  assertEquals(validResult, true, "Rendering with a registered function should return true");
});

Deno.test("TestModelFromObject", () => {
  const model = new Model();
  const data = {
    id: "testModel",
    description: "A test model",
    attributes: { attr1: "value1" },
    elements: [
      { id: "element1", type: "text", attributes: { name: "element1" }, isObjectId: true },
      { id: "element2", type: "textarea", attributes: { name: "element2" } },
    ],
  };

  const result = model.fromObject(data);
  assertEquals(result, true, "fromObject should return true for valid data");
  assertEquals(model.id, "testModel", "Model ID should be 'testModel'");
  assertEquals(model.description, "A test model", "Model description should be 'A test model'");
  assertEquals(model.attributes["attr1"], "value1", "Model attribute 'attr1' should be 'value1'");
  assertEquals(model.elements.length, 2, "Model should have 2 elements");
  assertEquals(model.elements[0].id, "element1", "First element ID should be 'element1'");
  assertEquals(model.elements[1].id, "element2", "Second element ID should be 'element2'");
});

Deno.test("TestYAMLConversion", () => {
  const model = new Model();
  const yamlDocument = `
  id: testModel
  description: ... description of testModel goes here ...
  attributes: {}
  elements:
    - id: key
      type: text
      attributes:
        name: key
        required: true
      is_object_id: true
    - id: textInput
      type: text
      attributes:
        name: textInput
        required: "true"
    - id: textArea
      type: textarea
      attributes:
        name: textArea
        required: "true"
    - id: submitButton
      type: submit
      attributes:
        value: Submit
  `;

  const obj = yaml.parse(yamlDocument) as { [key: string]: any };
  const result = model.fromObject(obj);
  assertEquals(result, true, "fromObject should return true for valid YAML data");
  assertEquals(model.check(), true, `expected true, got false for "${model.errors.join(', ')}"`);
});

Deno.test("TestModelBuilding", async () => {
  const modelId = "test_model";
  const model = Model.newModel(modelId);
  assert(model !== null, `Failed to create new model ${modelId}`);
  model.changed(false);
  assert(!model.hasChanges(), `${modelId} should not have changes yet`);

  const buf = new WritableStream();
  assert(model.check(), "Expected a valid model");
});

Deno.test("TestHelperFuncs", () => {
  const m = {
    one: "1",
    two: "2",
    three: "3",
  };
  const attrNames = ["one", "two", "three"];
  const got = Object.keys(m);
  assertEquals(got.length, 3, "Expected 3 attribute ids");
  for (const expected of attrNames) {
    assert(got.includes(expected), `Expected ${expected} in ${got}`);
  }
});

Deno.test("TestValidateModel", async () => {
  const src = `
id: test_validator
description: This is a test of the validation code
elements:
  - id: pid
    type: text
    attributes:
      name: pid
      required: true
    is_primary_id: true
    label: Personal Identifier
  - id: lived
    type: text
    attributes:
      name: lived
      required: true
    label: Lived Name
  - id: family
    type: text
    attributes:
      name: family
      required: true
    label: Family Name
  - id: orcid
    type: text
    pattern: "[0-9]{4}-[0-9]{4}-[0-9]{4}-[0-9]{3}[0-9A-Z]"
    attributes:
      name: orcid
      required: true
    label: ORCID
`;

  const model = Model.newModel("test_model");
  assert(model !== null, "Failed to create new model");
  const yamlData = await yaml.parse(src) as { [key: string]: any };
  model.fromObject(yamlData);

  const formData = {
    pid: "jane-doe",
    lived: "Jane",
    family: "Doe",
    orcid: "0000-1111-2222-3333",
  };

  assert(model.validate(formData), "Form data failed to validate");
});

Deno.test("TestValidateMapInterface", async () => {
  const src = `
id: test_validate_map_interface
description: This is a test of the validation code
elements:
  - id: pid
    type: text
    attributes:
      name: pid
      required: true
    is_primary_id: true
    label: Personal Identifier
    generator: uuid
  - id: lived
    type: text
    attributes:
      name: lived
      required: true
    label: Lived Name
  - id: family
    type: text
    attributes:
      name: family
      required: true
    label: Family Name
  - id: orcid
    type: text
    pattern: "[0-9]{4}-[0-9]{4}-[0-9]{4}-[0-9]{3}[0-9A-Z]"
    attributes:
      name: orcid
      required: true
    label: ORCID
  - id: created
    type: datetime-local
    attributes:
      required: true
    label: created
    generator: created_timestamp
  - id: updated
    type: datetime-local
    attributes:
      required: true
    generator: current_timestamp
`;

  const model = Model.newModel("test_model");
  assert(model !== null, "Failed to create new model");
  const yamlData = await yaml.parse(src) as { [key: string]: any };
  model.fromObject(yamlData);

  const pid = crypto.randomUUID();
  const formData = {
    pid,
    lived: "Jane",
    family: "Doe",
    orcid: "0000-1111-2222-3333",
    created: "2024-10-03T12:40:00",
    updated: "2024-10-03 12:41:32",
  };

  assert(model.validateMapInterface(formData), "Form data failed to validate");

  const formData2 = {
    created: "2024-10-03T13:25:24-07:00",
    family: "Jetson",
    lived: "George",
    orcid: "1234-4321-1234-4321",
    pid: "0192540f-0806-7631-b08f-4ae5c4d37cca",
    updated: "2024-10-03T13:25:24-07:00",
  };

  assert(model.validateMapInterface(formData2), "Form data failed to validate");
});

Deno.test("TestModelElements", () => {
  const model = new Model();
  const modelTypes = model.getGeneratedTypes();
  assertEquals(Object.keys(modelTypes).length, 0, "Expected zero model types");
});
